diff --git a/dist/cjs/rx-change-event.js b/dist/cjs/rx-change-event.js
index 73a80a4de09a57a58750d8953027117714c0f2d0..69541f793d24f9642eac5d1d387f36e97a28d9ce 100644
--- a/dist/cjs/rx-change-event.js
+++ b/dist/cjs/rx-change-event.js
@@ -87,7 +87,7 @@ function flattenEvents(input) {
   });
   return nonDuplicate;
 }
-var EVENT_BULK_CACHE = new Map();
+var EVENT_BULK_CACHE = new WeakMap();
 function rxChangeEventBulkToRxChangeEvents(eventBulk) {
   return (0, _index.getFromMapOrCreate)(EVENT_BULK_CACHE, eventBulk, () => {
     var events = new Array(eventBulk.events.length);
diff --git a/dist/cjs/rx-storage-helper.js b/dist/cjs/rx-storage-helper.js
index 32b8007f4ef3b71d392b8f62fe1bc0d989fb1ca3..0d918ce6a8534722f7a08213ea0cdeb89cdd831c 100644
--- a/dist/cjs/rx-storage-helper.js
+++ b/dist/cjs/rx-storage-helper.js
@@ -624,7 +624,7 @@ rxJsonSchema) {
       // No need to check writeResult.error.length === 0 here because
       // the fast path above already returns early when there are no errors.
       var reInsertErrors = writeResult.error.filter(error => {
-        if (error.status === 409 && !error.writeRow.previous && !error.writeRow.document._deleted && (0, _index4.ensureNotFalsy)(error.documentInDb)._deleted) {
+        if (context !== "studio-projection-pull" && error.status === 409 && !error.writeRow.previous && !error.writeRow.document._deleted && (0, _index4.ensureNotFalsy)(error.documentInDb)._deleted) {
           return true;
         }
 
diff --git a/dist/esm/rx-change-event.js b/dist/esm/rx-change-event.js
index 0063a9ad8d09b871d66a8b2c2df2abd6f8dea739..b7c43f524e4ed48fc5efcb6e8b188fb42ca59938 100644
--- a/dist/esm/rx-change-event.js
+++ b/dist/esm/rx-change-event.js
@@ -78,7 +78,7 @@ export function flattenEvents(input) {
   });
   return nonDuplicate;
 }
-var EVENT_BULK_CACHE = new Map();
+var EVENT_BULK_CACHE = new WeakMap();
 export function rxChangeEventBulkToRxChangeEvents(eventBulk) {
   return getFromMapOrCreate(EVENT_BULK_CACHE, eventBulk, () => {
     var events = new Array(eventBulk.events.length);
diff --git a/dist/esm/rx-storage-helper.js b/dist/esm/rx-storage-helper.js
index 9a2061f27464dc10164940c9efbe6e0b195c88c1..eaad0283892d82b43863e84f7072cdd48a3c2e18 100644
--- a/dist/esm/rx-storage-helper.js
+++ b/dist/esm/rx-storage-helper.js
@@ -600,7 +600,7 @@ rxJsonSchema) {
       // No need to check writeResult.error.length === 0 here because
       // the fast path above already returns early when there are no errors.
       var reInsertErrors = writeResult.error.filter(error => {
-        if (error.status === 409 && !error.writeRow.previous && !error.writeRow.document._deleted && ensureNotFalsy(error.documentInDb)._deleted) {
+        if (context !== "studio-projection-pull" && error.status === 409 && !error.writeRow.previous && !error.writeRow.document._deleted && ensureNotFalsy(error.documentInDb)._deleted) {
           return true;
         }
 
